Added ability to switch between points and pixels

vedranio opened the pull request:
– Added UI control for switching between points and pixels
– Preview now updates correctly taking unit of measure into account
– Default state now uses 18px as a starting point, instead of 14pt
– Added relevant tests for accessibleContrast

vedranio committed:
Added ability to switch between points and pixels

– Added UI control for switching between points and pixels
– Preview now updates correctly taking unit of measure into account
– Default state now uses 18px as a starting point, instead of 14pt
– Added relevant tests for accessibleContrast

moroshko commented on src/actions/app.js:
```diff
@@ -2,6 +2,7 @@ export const UPDATE_TEXT_COLOR = 'UPDATE_TEXT_COLOR';
 export const BLUR_TEXT_COLOR = 'BLUR_TEXT_COLOR';
 export const UPDATE_FONT_SIZE = 'UPDATE_FONT_SIZE';
 export const BLUR_FONT_SIZE = 'BLUR_FONT_SIZE';
+export const UPDATE_FONT_UNIT_OF_MEASURE = 'UPDATE_FONT_UNIT_OF_MEASURE'
```
How about we change the terminology from "Font unit of measure" to "Font size unit"? It's shorter, and also directly correlates with "Font size". If you happy with this change, we'll need to do it across the whole PR, not just here.

vedranio replied:
OK, makes sense

moroshko commented on src/components/App/App.js:
```diff
@@ -15,21 +15,22 @@ function mapStateToProps(state) {
   const fontSizeValue = parseInt(fontSize.value, 10);
   const accessibleContrastRatio = isUserInputValid ?
-    accessibleContrast(accessibilityLevel, fontSizeValue, isFontBold) : null;
+    accessibleContrast(accessibilityLevel, fontSizeValue, isFontBold, fontUnitOfMeasure) : null;
```
Can we change the order of the last 3 args to: `fontSizeValue, fontUnitOfMeasure, isFontBold`? This is the order in the UI, so would be nice to keep it consistent.

vedranio replied:
I thought I did it everywhere – must have missed a spot

moroshko commented on src/constants.js:
```diff
@@ -2,4 +2,4 @@ export const DOMAIN = 'accessible-colors.com';
 export const REPO = 'moroshko/accessible-colors';
 export const MIN_FONT_SIZE = 10;
-export const MAX_FONT_SIZE = 21;
+export const MAX_FONT_SIZE = 28;
```
Now, 22pt and above are not reflected in the UI, as it's bounded by 28px.
Shall we change `MAX_FONT_SIZE` to 36?

Also, could you change the names `MIN_FONT_SIZE` and `MAX_FONT_SIZE` to be `MIN_FONT_SIZE_IN_PT` and `MAX_FONT_SIZE_IN_PT` (or `IN_PX`) to be more specific?

vedranio replied:
Agreed

moroshko commented on src/constants.js:
```diff
@@ -2,4 +2,4 @@ export const DOMAIN = 'accessible-colors.com';
 export const REPO = 'moroshko/accessible-colors';
 export const MIN_FONT_SIZE = 10;
-export const MAX_FONT_SIZE = 21;
+export const MAX_FONT_SIZE = 28;
```
If you already touching max font size, do you mind deciding what to do with https://github.com/moroshko/accessible-colors/issues/63? (either make the change, or close the issue)

moroshko commented on src/reducers/app.js:
```diff
@@ -15,7 +15,7 @@ const initialTextColor = '#747474';
 const initialTextColorHSL = str2hsl(initialTextColor);
 const initialState = {
-  githubStars: '31',
+  githubStars: '51',
```
Nice one, but it's already changed 😄

moroshko commented on src/utils/accessibility/accessibility.js:
```diff
@@ -3,20 +3,32 @@ function isLargeScale(fontSize, isFontBold) {
 }
 
-function accessibleContrast(accessibilityLevel, fontSize, isFontBold) {
+function fontSizeInPt(fontSize,fontUnitOfMeasure) {
```
There is a minor styling issue here. Usually we put space after the comma. 
Could you run `npm run lint` to and clean the issues reported? (you'll see a few more)

moroshko replied:
I hope to find time to integrate [Prettier](https://github.com/prettier/prettier) which auto formats the code, so we don't discuss styling issues in the future.

vedranio replied:
will do

moroshko commented on src/utils/accessibility/accessibility.test.js:
```diff
@@ -5,25 +5,33 @@ describe('Accessibility Utils', () => {
+        expect(accessibleContrast('AA', 18, true, 'pt')).to.equal(4.5);
+        expect(accessibleContrast('AA', 24, false, 'pt')).to.equal(4.5);
       });
     });
```
Looks like we have 2 failing tests now.
Could you run `npm test` and fix them?

moroshko commented on src/variables.less:
```diff
@@ -1,5 +1,5 @@
 @small-screen-width: 320px;
-@medium-screen-width: 510px;
+@medium-screen-width: 520px;
```
Curious to know why did you change this?

vedranio replied:
The input lines wrapped too early if I left it at 510px. The 'pt' toggle takes up more space. I tested what it looks like at different breakpoints as you resize the screen and it works well

moroshko commented on src/utils/accessibility/accessibility.js:
```diff
@@ -3,20 +3,32 @@ function isLargeScale(fontSize, isFontBold) {
+
+function fontSizeInPx(fontSize,fontUnitOfMeasure) {
+  // http://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-contrast.html#visual-audio-contrast-contrast-73-head
+  return (fontUnitOfMeasure === 'pt') ? Math.round(fontSize * 1.333) : fontSize;
```
This would be easier to digest, and would be more consistent with `fontSizeInPt()`:
```js
return (fontUnitOfMeasure === 'px') ? fontSize : Math.round(fontSize * 1.333);
```

moroshko replied:
How about we put this magical number in a constant at the top of this file:

```js
// 1pt = 1.333px according to http://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-contrast.html#visual-audio-contrast-contrast-73-head
const PIXELS_IN_POINT = 1.333;
```

To use it, we'll do `fontSize * PIXELS_IN_POINT` or `fontSize / PIXELS_IN_POINT`.

moroshko commented on src/utils/accessibility/accessibility.js:
```diff
@@ -3,20 +3,32 @@ function isLargeScale(fontSize, isFontBold) {
+  return (fontUnitOfMeasure === 'pt') ? fontSize : Math.round(fontSize * 0.75);
+}
+
+function fontSizeInPx(fontSize,fontUnitOfMeasure) {
```
Well done on putting this function in this file even though it's not used here.

moroshko commented on src/utils/accessibility/accessibility.js:
```diff
@@ -3,20 +3,32 @@ function isLargeScale(fontSize, isFontBold) {
 export {
-  accessibleContrast
+  accessibleContrast,
+  fontSizeInPx
```
Usually, when we export a function from a module, we want to test it.
Can we add tests for `fontSizeInPx`?

vedranio committed:
Addressing PR feedback from @moroshko

vedranio committed:
removing debug logging

moroshko commented on src/components/App/components/UserInput/UserInput.js:
```diff
@@ -68,15 +70,22 @@ function UserInput(props) {
-                  min: MIN_FONT_SIZE,
-                  max: MAX_FONT_SIZE,
+                  min: MIN_FONT_SIZE_IN_PX,
+                  max: MAX_FONT_SIZE_IN_PX,
```
Now, that we have `IN_PX`, this doesn't look right.
We made the variable name more clear, and it highlighted a logical issue.
The question I ask myself here is, why the `min` and `max` are `IN_PX`. Why not `IN_PT`?
Indeed, if you change the unit to `pt`, you'll be able to go up to 36pt, which, again, reaches the area where changing the font size is not reflected in the UI (27-36pt looks the same because they all above our max which is 36px).
Now we have an interaction issue :)
What do you think should happen when you set `36px`, and then toggle to `pt`?
Should it jump to `24pt` because it's our maximum? (equivalent to 36px)
This jump from 36 to 24 can feel weird, but at least we don't allow users to get to a bad state (>24pt).
What do you think?

vedranio replied:
Yes, I thought about this a lot last night.

I don't like the idea of automatically adjusting the size input when you change the unit at all. It doesn't feel intuitive.

I would propose leaving it like this – I think the preview not updating when you get to those larger sizes is the lesser of the two evils.

vedranio replied:
Another idea could be to show a message (where we currently show the error message) explaining why the preview is not changing. I'm not convinced that it's necessary. What do you think?

moroshko replied:
What if we don't constrain the font size in the Preview, so the preview font size could be in the range 8..48px (which is 36pt).




What if we don't constrain the font size in the Preview, so it could range 
from: 8px
to:

moroshko commented on src/utils/accessibility/accessibility.js:
```diff
@@ -1,22 +1,36 @@
+}
+
+function fontSizeInPx(fontSize, fontSizeUnit) {
+  // http://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-contrast.html#visual-audio-contrast-contrast-73-head
```
Maybe we can remove this comment, and the one in `fontSizeInPt()`, since we already have it at the top of the file?

moroshko commented on src/utils/accessibility/accessibility.js:
```diff
@@ -1,22 +1,36 @@
+
 // http://www.w3.org/TR/2008/REC-WCAG20-20081211/#larger-scaledef
 function isLargeScale(fontSize, isFontBold) {
   return isFontBold ? (fontSize >= 14) : (fontSize >= 18);
```
I got really confused here because it looked like we ignore the font size unit here.
Then, I noticed that `fontSize` actually means `fontSizeInPt` because we call `fontSizeInPt()` before calling `isLargeScale()`.
Can I suggest to add a `fontSizeUnit` argument here, and call `fontSizeInPt()` here rather than in `accessibleContrast()`? This will also ensure that if we call `isLargeScale()` from another place in the future, we don't need to remember to call `fontSizeInPt()` first.

moroshko commented on src/utils/accessibility/accessibility.test.js:
```diff
@@ -1,29 +1,53 @@
+    });
+
+    describe('fontSizeInPx()', () => {
+      describe('Should', () => {
```
I'd prefer to avoid this extra `describe` (the less nesting, the easier it is to read the code). Usually, if it's the only `describe` (like in this case), it can be avoided. Just add `should` in `it()`, like in the tests above (e.g. `"should round up"`).

moroshko commented on src/utils/accessibility/accessibility.test.js:
```diff
@@ -1,29 +1,53 @@
+          expect(fontSizeInPx(14, 'pt')).to.equal(19);
+        });
+
+        it('rounded down', () => {
```
rounded => round?
"should round down" sounds better

vedranio committed:
Addressing PR feedback from @moroshko

vedranio committed:
fixed nesting for fontSizeInPX() test

moroshko commented on src/constants.js:
```diff
@@ -1,5 +1,5 @@
-export const MIN_FONT_SIZE = 10;
-export const MAX_FONT_SIZE = 21;
+export const MIN_FONT_SIZE_IN_PX = 8;
+export const MAX_FONT_SIZE_IN_PX = 36;
```
Why did we change the max from 21px to 36px?

vedranio replied:
Good question. Why do we have constraints at all? As long as the preview handles large fonts  nicely (num lines to show), we shouldn't have a constraint at either end.

Remove the constraints?

moroshko replied:
Yeah, let's remove them. Maybe we'll find out later why did we add them in the first place, but maybe it was just an overlook.
Could you also see if we need to add more lines in the Preview for 8px?

moroshko commented on src/utils/accessibility/accessibility.js:
```diff
@@ -2,29 +2,27 @@
-function isLargeScale(fontSize, isFontBold) {
-  return isFontBold ? (fontSize >= 14) : (fontSize >= 18);
+function isLargeScale(fontSize, fontSizeUnit, isFontBold) {
+  return isFontBold ? (fontSizeInPt(fontSize, fontSizeUnit) >= 14) : (fontSizeInPt(fontSize, fontSizeUnit) >= 18);
```
Let's DRY the code a bit:
```js
const points = fontSizeInPt(fontSize, fontSizeUnit);

return isFontBold ? (points >= 14) : (points >= 18);
```

vedranio committed:
Implementing PR feedback from @moroshko

– removed upper constraint for font size
– updated githubStars
– more efficient isLargeScale()
– updated Lipsum lines for Preview

moroshko commented on src/components/App/components/Preview/Preview.js:
```diff
@@ -2,23 +2,27 @@ import styles from './Preview.less';
   return Math.round(100 * contrast(color1, color2)) / 100;
 }
 
 function calcLinesToShow(fontSize) {
```
Lets call the parameter here `fontSizeInPx` to be explicit.

vedranio replied:
I'm guessing you mean throughout the file? Wherever the unit is in px, change the variable name?

moroshko replied:
Nope. I was referring to the parameter of the function `calcLinesToShow()`. So, just this function, not the whole file.

moroshko commented on src/components/App/components/UserInput/UserInput.js:
```diff
@@ -71,7 +71,6 @@ function UserInput(props) {
                   id: 'font-size',
                   type: 'number',
                   min: MIN_FONT_SIZE_IN_PX,
-                  max: MAX_FONT_SIZE_IN_PX,
```
Oh no! I didn't mean to remove the max here. I thought we are removing it only from the Preview?

vedranio replied:
Can you please explain what behaviour you're trying to achieve. I'm confused what the intent is if we're bounding the input, but not the Preview

moroshko replied:
Sure!

First, note that there is a difference between constraining the font size input and the Preview.

In the font size input, we are constraining a number, regardless of the font size unit. For example, if the max is 36, it means that user can get to 36px or 36pt, but not 37px or 37pt.

In the Preview, we are constraining using a unit. For example, if the max is 36px, it means that we never show text that is 37px. Even if user puts 36pt, which is 48px, we'll cut the font size down to the max of 36px.

By default, I'd say let's not constrain either.
The problem with this is, that users could get to 3 digit font sizes which we don't handle well (input is too narrow). So, it doesn't look good, and also I think it's not very useful. This is why I suggest we do constrain the user input.

Constraining user input means that the Preview will be constrained as well. We don't need to write code to constrain it! We just need to note that, if font size input is constrained by 36, for example, then the Preview will be automatically constrained by 48px (so the max number is not quite the same in both).

If this makes sense to you, I suggest we bring back `MAX_FONT_SIZE` and add a comment:
```js
export const MAX_FONT_SIZE = 21; // Maximum font size input value
```
(and let's add a similar comment for `MIN_FONT_SIZE`)

I think adding `IN_PX` only caused a lot of confusion.

vedranio replied:
Got ya. That makes sense – I misunderstood the earlier comment

moroshko commented on src/constants.js:
```diff
@@ -2,4 +2,3 @@ export const DOMAIN = 'accessible-colors.com';
 export const SITE_URL = `http://${DOMAIN}`;
 export const REPO = 'moroshko/accessible-colors';
 export const MIN_FONT_SIZE_IN_PX = 8;
-export const MAX_FONT_SIZE_IN_PX = 36;
```
Can we leave the max for the input to be bounded? Why did we increase it to 36px?

vedranio replied:
The max size we choose will be somewhat arbitrary. Once I make the other changes you commented on above, I will choose the largest number that looks good in the preview (when input is in pt).

moroshko commented on src/utils/accessibility/accessibility.js:
```diff
@@ -3,7 +3,8 @@ const PIXELS_IN_POINT = 1.333;
 function isLargeScale(fontSize, fontSizeUnit, isFontBold) {
-  return isFontBold ? (fontSizeInPt(fontSize, fontSizeUnit) >= 14) : (fontSizeInPt(fontSize, fontSizeUnit) >= 18);
+  const points = fontSizeInPt(fontSize, fontSizeUnit);
+  return isFontBold ? (points >= 14) : (points >= 18);
```
Make sure to run `npm run build` before committing. This will run the linter and the tests. (Now the build is broken because of a lint issue.)

vedranio committed:
Addressing PR comments from @moroshko

– put back the max constraint on input
– updated max font input size to 22

moroshko commented:
Great work @vedranio 👍
